Midend pass to detect entries with duplicate or unmatchable keys

ChrisDodd opened the pull request:
- duplicate keys can be either errors or warnings (per arch)
- unmatchable ternary keys (due to prior key that covers it) are warnings

jafingerhut commented on midend/checkTableEntries.cpp:
```diff
@@ -0,0 +1,106 @@
+    }
+}
+
+bool P4::CheckTableEntries::ternary_covers(const IR::Expression *k1, const IR::Expression *k2) {
```
For future readers, a comment like the following on `ternary_covers` might be helpful:
```
// Return true if the set of lookup keys matched by k1 is a superset
// of the keys matched by k2, including if they match the same set of lookup keys.
// Otherwise return false.
```

jafingerhut reviewed (approved):
LGTM

jafingerhut reviewed (changes requested):
Actually, before I approve, a question:

If a developer uses explicit numeric priorities on entries, then it seems correct to do the checks added by this PR only on the list of entries after they have been sorted from highest priority to lowest priority (and then, correctly handling the two cases of `largest_priority_wins = true` and `largest_priority_wins = false`.

Does this PR have code for sorting the entries by priority, if explicit priorities are specified?

ChrisDodd commented:
> If a developer uses explicit numeric priorities on entries, then it seems correct to do the checks added by this PR only on the list of entries after they have been sorted from highest priority to lowest priority (and then, correctly handling the two cases of `largest_priority_wins = true` and `largest_priority_wins = false`.

The exiting `EntryPriorities` frontend pass looks at explicit priorities on table entries and issues a warning if they are not in order.  Perhaps there should be an option there to sort them if they are out of order?

jafingerhut commented:
> > If a developer uses explicit numeric priorities on entries, then it seems correct to do the checks added by this PR only on the list of entries after they have been sorted from highest priority to lowest priority (and then, correctly handling the two cases of `largest_priority_wins = true` and `largest_priority_wins = false`.
> 
> The exiting `EntryPriorities` frontend pass looks at explicit priorities on table entries and issues a warning if they are not in order. Perhaps there should be an option there to sort them if they are out of order?

Before the check being added by this PR, I cannot think of any reason _not_ to sort them by priority order before performing the check.  NOT doing so seems like it could only lead to incorrect warnings (either incorrect issued warnings, or lack of a warning when one should have been issued).

ChrisDodd commented:
I added code to EntryPriorities to sort the entries into priority order when they are out of order (so subsequent passes like CheckTableEntries can assume they are in order).  I also removed the code that was incorrectly adding priorities to exact match tables (where they should not be -- their presence was generally being ignored by subsequent passes, so this error wasn't noticed)
